File: go/game/tile.go
package game

type Tileish interface {
	tile() *Tile
	dist(tileish Tileish) Distance
	stepOn(p Platform, s *State, monster Monstrous) error
	setEffect(sprite SpriteIndex)
}

type Tile struct {
	monster       Monstrous
	x, y          Position
	sprite        SpriteIndex
	passable      bool
	treasure      bool
	effect        SpriteIndex
	effectCounter counter
}

func NewTile(sprite SpriteIndex, x, y Position, passable bool) *Tile {
	return &Tile{
		x: x, y: y,
		sprite:   sprite,
		passable: passable,
	}
}

func (t *Tile) dist(tileish Tileish) Distance {
	tile := tileish.tile()

	return abs(Distance(t.x)-Distance(tile.x)) + abs(Distance(t.y)-Distance(tile.y))
}

// passes the minimum value (-2^N) through unchanged.
func abs(d Distance) Distance {
	if d < 0 {
		return -d
	}

	return d
}

const (
	effectDuration = 30
)

func (t *Tile) draw(p Platform, shake shake) {
	sprite(p, t.sprite, t.x, t.y, shake)

	if t.treasure {
		sprite(p, 12, t.x, t.y, shake)
	}

	if t.effectCounter.isActive() {
		t.effectCounter.dec()

		spriteWithAlpha(
			p,
			t.effect,
			t.x,
			t.y,
			shake,
			Alpha(float32(t.effectCounter.value)/effectDuration),
		)
	}
}

func (t *Tile) setEffect(sprite SpriteIndex) {
	t.effect = sprite
	t.effectCounter = counter{effectDuration}
}

func (t *Tile) tile() *Tile {
	return t
}

func (t *Tile) stepOn(_ Platform, s *State, monster Monstrous) (err error) {
	// Empty default implementation
	return
}

type TileMaker = func(x, y Position) Tileish

type Floor struct {
	*Tile
}

func NewFloor(x, y Position) Tileish {
	return &Floor{
		Tile: NewTile(2, x, y, true),
	}
}

func (t *Floor) stepOn(p Platform, s *State, monster Monstrous) (err error) {
	if _, isPlayer := monster.(*Player); isPlayer && t.treasure {
		s.score++

		if s.score%3 == 0 && s.numSpells < 9 {
			s.numSpells++
			s.player.addSpell(s.spells)
		}

		p.Sound(Treasure)

		t.treasure = false

		m, err := spawnMonster(s)
		if err != nil {
			return err
		}
		s.monsters = append(s.monsters, m)
	}

	return
}

type Wall struct {
	*Tile
}

func NewWall(x, y Position) Tileish {
	return &Wall{
		Tile: NewTile(3, x, y, false),
	}
}

type Exit struct {
	*Tile
}

func NewExit(x, y Position) Tileish {
	return &Exit{
		Tile: NewTile(11, x, y, true),
	}
}

func (t *Exit) stepOn(p Platform, s *State, monster Monstrous) (err error) {
	_, isPlayer := monster.(*Player)

	if isPlayer {
		p.Sound(NewLevel)

		if s.level == numLevels {
			addScore(p, s.score, Won)
			s.state = title
		} else {
			s.level++

			newHP := s.player.hp + 1
			if newHP > maxHP {
				newHP = maxHP
			}

			err = startLevel(s, newHP, nil)
		}
	}

	return
}

File: nim/tile.nim
from options import Option, none, isSome

from game import no_ex, Shake, Counter, `<`
from monster import Monster

const EffectCounterMax = 255

type
  Kind* = enum
    Wall,
    Floor,
    Exit

  Effect* = tuple
    sprite: game.SpriteIndex
    counter: Counter

  Tile* = tuple
    kind: Kind
    xy: game.TileXY
    monster: Option[Monster]
    treasure: bool
    effect: Effect

no_ex:
    func newTile(kind: Kind, xy: game.TileXY): Tile =
        (kind: kind, xy: xy, monster: none(Monster), treasure: false, effect: (sprite: game.SpriteIndex(1), counter: Counter(0)))
    
    func newWall*(xy: game.TileXY): Tile =
        newTile(Kind.Wall, xy)

    func newFloor*(xy: game.TileXY): Tile =
        newTile(Kind.Floor, xy)

    func newExit*(xy: game.TileXY): Tile =
        newTile(Kind.Exit, xy)
    
    proc draw*(tile: var Tile, shake: Shake, platform: game.Platform) =
        let sprite = case tile.kind
        of Wall:
            game.SpriteIndex(3)
        of Floor:
            game.SpriteIndex(2)
        of Exit:
            game.SpriteIndex(11)

        let floatXY = (
            x: float(tile.xy.x),
            y: float(tile.xy.y)
        )
    
        (platform.spriteFloat)(
            shake,
            sprite,
            floatXY
        )

        if tile.treasure:
            (platform.spriteFloat)(
                shake,
                game.SpriteIndex(12),
                floatXY
            )

        if tile.effect.counter > 0:
            (platform.spriteFloat)(
                shake,
                tile.effect.sprite,
                floatXY,
                float(tile.effect.counter) / EffectCounterMax
            )
            tile.effect.counter.dec()
            
    proc setEffect*(tile: var Tile, sprite: game.SpriteIndex) =
        tile.effect = (sprite: sprite, counter: Counter(EffectCounterMax))

    func isPassable*(kind: Kind): bool =
        case kind
        of Wall:
            false
        else:
            true
        
    func isPassable*(t: Tile): bool =
        t.kind.isPassable

        
    func hasMonster*(t: Tile): bool =
            t.monster.isSome
